torch/csrc/PyInterpreter.cpp

```diff
@@ -565,6 +565,149 @@ static void set_tensor_attr_with_capsule(
+  // tensor that are <= 5 elements will not reallocate the buffer. But once we
+  // hit a size that is > 5 elements, we will re-allocate on every future
+  // resize. NOTE: Ideally, we shouldn't need to model the SmallVector
+  // optimization! But removing this optimization fails tests: there's some code
+  // in our codebase that relies on the fact that calling `.sizes()` and then
+  // resizing the tensor doesn't reallocate the underlying SmallVector. Ideally,
+  // we should kill this optimization and fix any places in our code that rely
+  // on this. Example failing test: test/functorch/test_aotdispatch.py
+  // TestPartitioning
+  bool needs_resize = false;
+  // We need to resize if:
+  // (1) we haven't allocated our buffer at all yet
+  // (2) our buffer is size 5 (smallvector optimization), but our new size is >
+  // 5 (3) our buffer is no longer size 5 (because at some point previously we
+  // saw a size > 5),
+  //     and our new size is different from the current buffer size
+  needs_resize = !is_buffer_allocated ||
+      ((!is_smallvector || new_size > 5) && curr_size != new_size);
+  if (needs_resize) {
+    auto set_to_smallvector = false;
+    // If our current buffer is not the right size (eithr because we haven't
+    // allocated it yet, or there was a metadata mutation that changed the
+    // number of dims of the tensor), allocate a fresh buffer. Note that this
+    // will trash the previous buffer if there already was one, invalidating any
+    // existing SymIntArrayRef's from an old .sym_size() call.
+    auto new_buffer_size = new_size;
+    if (!is_buffer_allocated && new_size <= 5) {
+      // This is the smallvector optimization
+      new_buffer_size = 5;
+      set_to_smallvector = true;
+    }
+    T* ptr = new T[new_buffer_size];
+    auto capsule =
+        py::capsule(ptr, [](void* p) { delete[] reinterpret_cast<T*>(p); });
+    int64_t idx = 0;
+    for (auto it = obj.begin(); it != obj.end(); ++it, ++idx) {
+      ptr[idx] = py::cast<T>(*it);
+    }
+    // Set the buffer
+    set_tensor_attr_with_capsule(tensor, capsule, base_attr_name);
+    // Set the len buffer
+    py::handle(mb_obj.value()).attr(buffer_len_attr_name.c_str()) = new_size;
+    // Set whether or not the buffer is now using the smallvector optimization
+    // (sized to 5)
+    py::handle(mb_obj.value()).attr(buffer_issmallvector_attr_name.c_str()) =
+        set_to_smallvector;
+  } else {
+    auto curr_buffer_pyobj = py::handle(mb_obj.value()).attr(base_attr_name);
+    void* buffer_pycapsule =
+        PyCapsule_GetPointer(curr_buffer_pyobj.ptr(), nullptr);
+    auto curr_buffer = reinterpret_cast<T*>(buffer_pycapsule);
+
+    // Overwrite the buffer with our new values, but only if any of them changed
+    // (due to a metadata mutation).
+    // This is technically not thread safe (maybe we should put the GIL here?),
+    // but *only* if we actually have to update, which only happens if there's a
+    // metadata mutation.
```

Comment:
This is not quite accurate because the update here can be delayed. So it could be that the original metadata mutation was thread safe, but because we lazily update, we don't end up mutating until we're out of the critical region and there is possibility for data race.

Reply:
to confirm: do you think this is something we should try to fix?

Agreed that the comment is inaccurate, I'll fix (you're pointing out that it could be the case that a metadata mutation happened in eager mode in a thread safe way, but our lazy update ends up not being thread safe)

Reply:
It's kind of hard to fix, because you have to do the update lazily. Having a lock is probably the only way to do it. Comment update is OK for this PR.